docs(getting-started): add routing page (#7495)
nuxt/framework: docs/content/1.getting-started/5.routing.md

# Routing

> One of Nuxt core feature is the file-system router. Every Vue file created inside the pages/ directory creates a corresponding URL (or route) that displays the content of the file.

Reviewer: ```suggestion
One of Nuxt core feature is the file-system router. Every Vue file created inside the pages/ directory creates a corresponding URL (or route) that displays the content of the file. It also code-split every page to a separate chunk by leveraging dynamic imports to ship the minimum of JavaScript for the requested route.
```

## Pages

> Nuxt routing is based on Vue-router. Under the hood, Nuxt generates the vue-router configuration file with every components created in the pages directory, based on their file name.

Reviewer 2: Would be nice to have more external links for readers.
```suggestion
Nuxt routing is based on [vue-router](https://router.vuejs.org/). Under the hood, Nuxt generates the vue-router configuration file with every components created in the pages directory, based on their file name.
```

> Nuxt routing is based on [vue-router](https://router.vuejs.org/). Under the hood, Nuxt generates the vue-router configuration file with every components created in the pages directory, based on their file name.

Reviewer: ```suggestion
Nuxt routing is based on [vue-router](https://router.vuejs.org/) and generates the routes configuration with every components created in the [`pages/`](/guide/directory-structure/pages) directory, based on their file name.
```

This file system routing uses naming conventions to create dynamic and nested routes:

::code-group

```text [pages/ directory]
pages/
--| about.vue
--| posts/
----| [id].vue
```

```js [Generated Router file]
{
  "routes": [
    {
      "path": "/about",
      "component": "pages/about.vue"
    },
    {
      "path": "/posts/:id",
      "component": "pages/posts/[id].vue"
    }
  ]
}
```

::

## Navigation

> The `<NuxtLink>` component links pages between them. It renders a `<a>` tag with the `href` attribute set to the route of the page. Once the application is hydrated, pages transitions are performed in JavaScript by updating the browser URL. This prevents full-page refreshes and allow for animated transitions.

Reviewer 2: We _could_ introduce prefetching feature too! It just landed for RC.10

Reviewer: Happy to make another PR once RC10 is landed?

```vue [pages/app.vue]
<template>
  <header>
    <nav>
      <ul>
        <li><NuxtLink to="/about">About</NuxtLink></li>
        <li><NuxtLink to="/posts/1">Post 1</NuxtLink></li>
        <li><NuxtLink to="/posts/2">Post 2</NuxtLink></li>
      </ul>
    </nav>
  </header>
</template>
```

:ReadMore{link="/api/components/nuxt-link"}

## Route Parameters

The `useRoute()` composable can be used in a `<script setup>` block or a `setup()` method of a Vue component to access the current route details.

```vue [pages/post/[id].vue]
<script setup>
const route = useRoute()

// When accessing /posts/1, route.params.id will be 1
console.log(route.params.id)
</script>
```

> 

Reviewer: ```suggestion
:ReadMore{link="/api/composables/use-route"}

```

## Route Middleware

Nuxt provides a customizable route middleware framework you can use throughout your application, ideal for extracting code that you want to run before navigating to a particular route.

::alert{type=info}
Route middleware run within the Vue part of your Nuxt app. Despite the similar name, they are completely different from server middleware, which are run in the Nitro server part of your app.
::

There are three kinds of route middleware:

1. Anonymous (or inline) route middleware, which are defined directly in the pages where they are used.
2. Named route middleware, which are placed in the `middleware/` directory and will be automatically loaded via asynchronous import when used on a page. (**Note**: The route middleware name is normalized to kebab-case, so `someMiddleware` becomes `some-middleware`.)

> 3. Global route middleware, which are placed in the `middleware/` directory (with a `.global` suffix) and will be automatically run on every route change.

Reviewer 2: Maybe add an example for middleware?

The `useRoute()` composable can be used in a `<script setup>` block or a `setup()` method of a Vue component to access the current route details.

```vue [pages/post/[id].vue]
<script setup>
const route = useRoute()

// When accessing /posts/1, route.params.id will be 1
console.log(route.params.id)
</script>
```
:ReadMore{link="/api/composables/use-route"}

## Route Middleware

Nuxt provides a customizable route middleware framework you can use throughout your application, ideal for extracting code that you want to run before navigating to a particular route.

::alert{type=info}
Route middleware run within the Vue part of your Nuxt app. Despite the similar name, they are completely different from server middleware, which are run in the Nitro server part of your app.
::

There are three kinds of route middleware:

1. Anonymous (or inline) route middleware, which are defined directly in the pages where they are used.
2. Named route middleware, which are placed in the `middleware/` directory and will be automatically loaded via asynchronous import when used on a page. (**Note**: The route middleware name is normalized to kebab-case, so `someMiddleware` becomes `some-middleware`.)

> 3. Global route middleware, which are placed in the `middleware/` directory (with a `.global` suffix) and will be automatically run on every route change.

Reviewer: ```suggestion
3. Global route middleware, which are placed in the `middleware/` directory (with a `.global` suffix) and will be automatically run on every route change.

Example of an `auth` middleware protecting the `/dashboard` page:

::code-group

```ts [middleware/auth.ts]
export default defineNuxtRouteMiddleware((to, from) => {
  if (isAuthenticated() === false) {
    return navigateTo('/login')
  }
})
```

```html [pages/dashboard.vue]
<script setup>
definePageMeta({
  middleware: 'auth'
})
</script>

<template>
  <h1>Welcome to your dashboard</h1>
</template>
```

::

:ReadMore{link="/guide/directory-structure/middleware"}
```
